src/vdaf/mastic.rs

```diff
@@ -0,0 +1,473 @@
+                nonce,
+            )?
+            .share
+            + self
+                .vidpf()
+                .eval(
+                    &vidpf_keys[1],
+                    &public_share,
+                    &VidpfInput::from_bools(&[true]),
+                    nonce,
+                )?
+                .share;
+
+        let szk_proof_shares = self.szk.prove(
+            leader_measurement_share.as_slice(),
+            helper_measurement_share.as_slice(),
+            measurement_weight.as_slice(),
+            szk_random,
+            opt_random,
+            nonce,
+        )?;
+        let leader_share = MasticInputShare::<T::Field, SEED_SIZE> {
+            vidpf_key: vidpf_keys[0].clone(),
+            proofs_share: szk_proof_shares[0].clone(),
+        };
+        let helper_share = MasticInputShare::<T::Field, SEED_SIZE> {
+            vidpf_key: vidpf_keys[1].clone(),
+            proofs_share: szk_proof_shares[1].clone(),
+        };
+        Ok((public_share, vec![leader_share, helper_share]))
+    }
+
+    pub(crate) fn is_randomized(&self) -> bool {
+        self.szk.has_joint_rand()
+    }
+
+    pub(crate) fn proof_len(&self) -> usize {
+        self.szk.proof_len()
+    }
+
+    fn encode_measurement(
+        &self,
+        measurement: &T::Measurement,
+    ) -> Result<VidpfWeight<T::Field>, VdafError> {
+        Ok(VidpfWeight::<T::Field>::from(
+            self.szk.typ.encode_measurement(measurement)?,
+        ))
+    }
+}
+
+impl<T, P, const SEED_SIZE: usize> Client<16> for Mastic<T, P, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    fn shard(
+        &self,
+        measurement: &(VidpfInput, T::Measurement),
+        nonce: &[u8; 16],
+    ) -> Result<(Self::PublicShare, Vec<Self::InputShare>), VdafError> {
+        if measurement.0.len() != self.bits {
+            return Err(VdafError::Vidpf(VidpfError::InvalidLabelLength));
+        }
+
+        let vidpf_keys = [
+            VidpfKey::gen(VidpfServerId::S0)?,
+            VidpfKey::gen(VidpfServerId::S1)?,
+        ];
+        let opt_random = match self.szk.has_joint_rand() {
+            true => Some(Seed::<SEED_SIZE>::generate()?),
+            false => None,
+        };
+        let szk_random = [
+            Seed::<SEED_SIZE>::generate()?,
+            Seed::<SEED_SIZE>::generate()?,
+        ];
+        self.shard_with_random(
+            &measurement.0,
+            &self.encode_measurement(&measurement.1)?,
+            nonce,
+            vidpf_keys,
+            szk_random,
+            opt_random,
+        )
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::field::Field128;
+    use crate::flp::types::Sum;
+    use rand::{thread_rng, Rng};
+
+    const TEST_WEIGHT_LEN: usize = 5;
+    const TEST_NONCE_SIZE: usize = 16;
+
+    #[test]
+    fn test_mastic_shard_sum() {
+        let algorithm_id = 6;
+        let sum_typ = Sum::<Field128>::new(5).unwrap();
```

Comment:
We should have some tests using a circuit that uses joint randomness as well, for better coverage